Modernize crypto library to v4.0.0 with enhanced Ed25519 support

bplatz opened the pull request:
## Summary

This PR modernizes the crypto library to support Verifiable Credentials, did:key standardization (moving away from did:fluree), and Ed25519 improvements that enable future multi-signature capabilities and other advanced cryptographic features.

- Migrate to @noble/ed25519 v2.3.0 for JavaScript environments (replaces SJCL)
- Implement Java 17+ native Ed25519 support using EdECPrivateKeySpec/EdECPublicKeySpec
- Add automatic public key derivation capabilities for simplified signing workflows
- Enhanced JWS implementation with flexible key identification options
- Convert to ESM-only JavaScript builds for better module compatibility
- Remove unused cryptographic algorithms (SECP256K1, SHA-3, RIPEMD, scrypt) no longer used by fluree/db v3
- Updated documentation with comprehensive API examples

## Test plan

- [x] All existing tests pass (JVM, Node.js, and browser environments)
- [x] Cross-platform consistency verified between CLJ and CLJS implementations
- [x] JWS roundtrip functionality tested with various key identification methods
- [x] Ed25519 compatibility verified across different key formats
- [x] GraalVM native image compilation confirmed working
- [x] Enhanced test coverage for new API functions

## Technical Details

### Breaking Changes
- **Requires Java 17+** for JVM environments (necessary for native Ed25519 support)
- ESM-only JavaScript builds (no CommonJS support)
- Removed unused cryptographic algorithms (SECP256K1, SHA-3, RIPEMD, scrypt) no longer used by fluree/db v3

### New Features
- Public key derivation from private key hex strings
- JWS signing with automatic key embedding options
- Enhanced account ID generation and DID support
- Improved GraalVM native image compatibility

### Dependencies
- @noble/ed25519 v2.3.0 (JavaScript environments)
- Java 17+ EdECPrivateKeySpec/EdECPublicKeySpec (JVM environments)
- Compatible with Clojure 1.12.1 and ClojureScript 1.11.132

Note: ClojureScript 1.12.42 compatibility requires shadow-cljs version updates that are not yet stable.

bplatz committed:
Remove SHA3/RIPEMD dependencies for GraalVM compatibility

Breaking changes for v4.0.0:

- Remove SHA3 and RIPEMD160 crypto functions and dependencies
- Update account ID generation to use SHA-256 truncation (Cosmos-style)
- Switch AES key derivation from SHA3-512 to SHA2-512
- Remove fluree.crypto.sha3 and fluree.crypto.ripemd namespaces
- Update JavaScript exports to remove deprecated functions
- Update tests to reflect new crypto implementation

These changes eliminate external crypto library dependencies
for SHA3 and RIPEMD160, enabling cleaner GraalVM native builds
using only Java standard library crypto primitives.

Account IDs will change due to algorithm switch - this is a
breaking change requiring coordination with consuming applications.

bplatz committed:
Upgrade to Java 17 for enhanced GraalVM support

- Add .java-version file specifying Java 17
- Update Dockerfile to use Java 17 base image (temurin-17)
- Verify all crypto operations work correctly with Java 17
- Confirm Java 17 standard library crypto support (SHA-256, SHA-512, AES)

Java 17 provides better GraalVM native image compatibility and
includes all required crypto primitives in the standard library,
reducing external dependencies for native builds.

bplatz committed:
Migrate to Noble crypto libraries - initial implementation

- Update dependencies from SJCL/js-sha3 to @noble/curves and @noble/hashes
- Replace secp256k1 operations with Noble curves implementation
- Replace HMAC operations with Noble hashes implementation
- Replace BigNumber operations with native BigInt (partial)
- Update package.json and deps.cljs to use Noble libraries
- Consolidate duplicate CLJ/CLJS code where platforms now work identically
- All Clojure tests passing (13 tests, 3022 assertions)

Work in progress:
- ClojureScript compilation has warnings for remaining SJCL references
- Some BigInt operations need platform-specific handling
- Signature recovery operations need Noble implementation
- Additional codec operations need updating

This provides foundation for completing GraalVM-compatible crypto operations.

bplatz committed:
Migrate from SJCL to Noble crypto libraries

BREAKING CHANGE: Signature format changed from custom DER+recovery to standard 65-byte format

- Replace @fluree/sjcl with @noble/curves and @noble/hashes
- Migrate secp256k1 operations to use @noble/curves/secp256k1
- Migrate HMAC operations to use @noble/hashes/hmac
- Replace SJCL BigNumber with native JavaScript BigInt
- Modernize signature recovery to use Noble's native API
- Change signature format to industry-standard 65-byte: r(32) + s(32) + recovery_id(1)
- Remove complex manual elliptic curve operations (~200 lines)
- Deprecate scrypt functions in ClojureScript (kept random-bytes)
- Fix all ClojureScript compilation issues

Note: Clojure implementation still needs updating for new signature format

bplatz committed:
Modernize crypto library to v4.0.0 with enhanced Ed25519 support

- Upgrade to Clojure 1.12.1 and Noble Ed25519 v2.3.0
- Remove core.async dependency in favor of direct API calls
- Add Java 17+ EdECPrivateKeySpec support for private key-only signing
- Implement comprehensive JWS functionality with key identification
- Add DID:key support and automatic public key derivation
- Convert to ESM-only JavaScript builds for better compatibility
- Add cross-platform consistency tests and GraalVM native image support
- Update documentation with Clojure and JavaScript examples
- Remove obsolete algorithms (SECP256K1, SHA-3, RIPEMD-160, Scrypt)
- Enhance test coverage with compatibility and integration tests

zonotope commented on resources/build/advanced/fluree-crypto.js:
```diff
@@ -0,0 +1,476 @@
+if(typeof Math.imul == "undefined" || (Math.imul(0xffffffff,5) == 0)) {
```
Should this build output be committed? I think all other things equal I would prefer compiling artifacts during a release process each time instead of checking build artifacts into versio control.

bplatz replied:
All cleaned out per your comment @zonotope and ready for further review if needed.

bplatz committed:
Add CommonJS build support alongside ESM

- Add :node-commonjs build target using :node-library
- Update package.json with dual module exports (ESM/CommonJS)
- Support both import and require for maximum compatibility
- Resolve ESM compilation issues with upstream libraries
- Maintain consistent naming conventions across build targets

bplatz committed:
Remove build artifacts and obsolete SJCL externs

- Add resources/build/ to .gitignore to prevent committing build artifacts
- Remove previously committed build artifacts from version control
- Remove obsolete sjcl-externs.js file since SJCL was replaced with Noble crypto libraries

bplatz committed:
Remove ClojureScript compilation artifacts from source control

- Add resources/public/js/out to .gitignore
- Remove 287 ClojureScript build output files from version control
- These files are generated during development and should not be tracked

zonotope reviewed (approved):
⌨️
